Add default layout option in order to use app/views/layouts

jaredcwhite opened the pull request:
### Summary

I was digging a ton into component previews on a new project today, and I found myself wanting to use a custom layout (in my case `component_preview`) for all of the index pages as well as component preview pages. I started by monkey patching Rails::ApplicationController, but then realized I should simply adjust what's going on in ViewComponentsController, which gave me the idea for a configuration option and hence this PR.

### Other Information

This makes no behavioral change at all out-of-the-box. Everything will work as before. The only difference is if that `default_preview_layout` config option is set, the preview indices and components will pull that in rather than use the Rails internal app layout. Of course individual previews can still override the layout, including setting it to nil. (I also ended up fixing an existing test that would always pass even if a layout _was_ rendered.)

Let me know what you think! Thanks.

jonspalmer commented on app/controllers/rails/view_components_controller.rb:
```diff
@@ -16,23 +16,31 @@ class Rails::ViewComponentsController < Rails::ApplicationController # :nodoc:
   def index
     @previews = ViewComponent::Preview.all
     @page_title = "Component Previews"
+    opts = default_preview_layout.nil? ? {} : { layout: default_preview_layout }
+    prepend_application_view_paths if opts.key?(:layout)
```
Can we set the layout explicitly rather than the indirection of `prepend_application_view_paths`.
```suggestion
    opts = default_preview_layout.nil? ? {} : { layout: Rails.root.join("app/views/layouts",  default_preview_layout) }
```

jaredcwhite replied:
@jonspalmer Well prepending the app's view path allows the layout to use partials from the app (shared/head, etc.), so that needs to be preserved.

jonspalmer replied:
Got it. And I'm guessing that sadly resolving those partial's isn't local to the layout that is picked. It relies of view paths?

jaredcwhite replied:
Let me see if I can refactor the layout hash stuff into a separate method so the individual actions in this controller are simpler.

jonspalmer commented on app/controllers/rails/view_components_controller.rb:
```diff
@@ -16,23 +16,31 @@ class Rails::ViewComponentsController < Rails::ApplicationController # :nodoc:
       @render_args = @preview.render_args(@example_name)
       layout = @render_args[:layout]
-      opts = layout.nil? ? {} : { layout: layout }
+      opts = if layout.nil?
+               default_preview_layout.nil? ? {} : { layout: default_preview_layout }
+             else
+               { layout: layout }
+             end
```
Functionally this is great but stylistically nested if conditions like this feel a little hard to read. It would be a little clearer as 
```suggestion
      layout_override = layout.presence? || default_preview_layout.presence?
      opts = layout_override.nil? ? {} : { layout: layout_override }
```

Do we have a test case where layout = false and default_preview_layout is set? Important that we're getting that right by checking nil? presence? etc

jonspalmer replied:
Similar comment as above about explicitly setting the layout rather than using `prepend_application_view_paths`

jaredcwhite replied:
> Functionally this is great but stylistically nested if conditions like this feel a little hard to read. It would be a little clearer as

I'm fine with cleaning that up a bit, but that syntax change didn't pass the a couple tests.

> Do we have a test case where layout = false and default_preview_layout is set? Important that we're getting that right by checking nil? presence? etc

Good call, adding a test to check false and switching logic from .nil? to .blank? did the trick.

jonspalmer commented on lib/view_component/previewable.rb:
```diff
@@ -7,6 +7,12 @@ module Previewable
     extend ActiveSupport::Concern
 
     included do
+      # Set a custom default preview layout through app configuration:
```
Can we add details that we're using this preview layout for the preview index pages too?

jonspalmer commented on test/view_component/integration_test.rb:
```diff
@@ -214,12 +215,28 @@ class IntegrationTest < ActionDispatch::IntegrationTest
+  test "test preview index renders custom application layout if configured" do
+    ViewComponent::Base.default_preview_layout = "admin"
+    get "/rails/view_components"
+    ViewComponent::Base.default_preview_layout = nil
```
For clarity on the code under test it feels like we're getting to a place where this would be considerably clearer as its own test class with setup and teardown methods that manipulate `default_preview_layout`. As it's written I did a double take a) we repeat this pattern in many tests b) is it important to nil out `default_preview_layout` before asserting or could doing that have side effects on the asserted conditions?


@joelhawksley what are you thoughts on using RSepc? The ability to use nested describe blocks would be perfect for cases like this where you would use before/after blocks to do common setup and teardown tasks.

jaredcwhite replied:
No strong opinions here about the test suite, so I'll happily defer to the rest of y'all.

juanmanuelramallo replied:
IMHO RSpec would be a great addition to the development environment of the gem @jonspalmer

joelhawksley replied:
How about we start by moving these tests to their own file?

I’d like to _try_ to stick with our current testing framework as it is consistent with Rails 😄

jaredcwhite replied:
@jonspalmer @joelhawksley If you need me to go rejigger the tests, just let me know

juanmanuelramallo replied:
In order not to disrupt the previous config I would suggest to set the value back instead of nil, we can go even a step further and create a helper method for this in the `test/test_helper.rb` file

```ruby
# test/test_helper.rb

...

def with_default_preview_layout(layout)
  old_value = ViewComponent::Base.default_preview_layout
  ViewComponent::Base.default_preview_layout = layout
  yield
  ViewComponent::Base.default_preview_layout = old_value
end
```
Then in our tests we can just use it like:
```ruby
test "test preview index renders custom application layout if configured" do
  with_default_preview_layout("admin") do
    get "/rails/view_components"
    assert_select "title", "ViewComponent - Admin - Test"
  end
end
```

jonspalmer commented on test/view_component/integration_test.rb:
```diff
@@ -214,12 +215,28 @@ class IntegrationTest < ActionDispatch::IntegrationTest
+    get "/rails/view_components"
+    ViewComponent::Base.default_preview_layout = nil
+
+    assert_select "title", "ViewComponent - Admin - Test"
```
Is it time for us to write a helper method for this pattern that makes it clearer we're asserting the right layout? That we're proxying the laying being applied by asserting the page title is sort of a weird detail. 
```suggestion
    assert_admin_layout
```

?

jonspalmer commented on test/view_component/integration_test.rb:
```diff
@@ -200,6 +200,7 @@ class IntegrationTest < ActionDispatch::IntegrationTest
     get "/rails/view_components/no_layout/default"
 
     assert_select("div", "hello,world!")
+    refute_includes response.body, "ViewComponent - Test"
```
Let's add a new test case for no_layout where we set `default_preview_layout` and make sure that there is still no layout.

joelhawksley replied:
@jaredcwhite thoughts on @jonspalmer's comment here?

jaredcwhite replied:
@joelhawksley I think I did that in a subsequent commit, but I'll double-check

jaredcwhite commented:
All right, I refactored the layout logic into its own method and merged the prepend views functionality into that as well. Still slightly non-obvious why the conditional is set up that way, but hopefully added comments make it clearer what to expect. Also added a test for the `layout false` scenario in a component.

juanmanuelramallo commented:
This PR will close https://github.com/github/view_component/issues/178

joelhawksley commented:
@jonspalmer might you be able to revisit and review this PR?

juanmanuelramallo commented:
I think we need to rebase this PR with master given that the view_components_controller was updated to use the templates from `app/views`. @jaredcwhite let me know if you need a hand with that

jaredcwhite commented:
@juanmanuelramallo Thanks for the suggestion regarding that test helper. I incorporated it into the tests and also moved them over to a separate file, after rebasing from master.

juanmanuelramallo commented on app/controllers/view_components_controller.rb:
```diff
@@ -16,24 +16,26 @@ class ViewComponentsController < Rails::ApplicationController # :nodoc:
   def index
     @previews = ViewComponent::Preview.all
     @page_title = "Component Previews"
+    render "view_components/index", **determine_layout
```
Will this work?
```suggestion
    render **determine_layout
```

juanmanuelramallo commented on app/controllers/view_components_controller.rb:
```diff
@@ -56,7 +58,26 @@ def set_locale
+      layout_declaration[:layout] = default_preview_layout
+    end
+
+    if layout_declaration.key?(:layout) || !!always_use_app_views
```
IMHO no need to double negate, it's just a boolean.

```suggestion
    if layout_declaration.key?(:layout) || always_use_app_views
```

juanmanuelramallo commented on test/view_component/custom_layouts_integration_test.rb:
```diff
@@ -0,0 +1,44 @@
+      refute_includes response.body, "ViewComponent - Admin - Test"
+    end
+  end
+end
```
Maybe we can name this file the same as the config attribute?

`test/view_component/default_preview_layout_integration_test.rb`

juanmanuelramallo commented on test/view_component/custom_layouts_integration_test.rb:
```diff
@@ -0,0 +1,44 @@
+require "test_helper"
+
+class CustomLayoutsIntegrationTest < ActionDispatch::IntegrationTest
+  test "test preview index renders custom application layout if configured" do
```
Maybe we can avoid rewriting "test" in each test case?

```suggestion
  test "preview index renders custom application layout if configured" do
```

jaredcwhite commented:
Good suggestions @juanmanuelramallo — I just pushed up those changes. Something weird happened with the CI though, nothing ran

juanmanuelramallo commented:
Weird, can you try rebasing and pushing so that the CI triggers again? I don't think it's related to your changes

jaredcwhite commented:
@juanmanuelramallo Working better now, but it looks like that one `render` change on the `index` action is running into a Rubocop rule:

`GitHub/RailsControllerRenderLiteral: render must be used with a string literal`

Should we disable it, or put back in the view name string?

juanmanuelramallo commented:
Oh man, yeah maybe put back the template name no? I'd rather do that than disabling rubocop for that line

jaredcwhite commented:
Works! ✅

jaredcwhite commented:
Hi all, just wanted to pop my head back in to see what the status might be for this feature. I haven't followed other PRs too closely lately so I'm not sure if this has been superseded by some other approach or if it's still relevant. Thanks =)

juanmanuelramallo commented:
Seems relevant to me, but same as you, haven't been following up other PRs closely. Better call @joelhawksley

jaredcwhite commented:
Sounds good. 👍 I hope to get to it this week.

jaredcwhite commented:
OK, at long last I was able to rebase and add some info to the README, so I think this is good to go now. Incidentally, there had been a regression on master regarding the context of `layout` set to `false` for a component which wasn't actually getting caught by a test (aka the test was passing even though a layout was in fact rendered), so this PR fixes that and makes sure the test catches any future regression.

juanmanuelramallo commented on app/controllers/view_components_controller.rb:
```diff
@@ -28,11 +28,11 @@ def previews
       locals = @render_args[:locals]
       opts = {}
-      opts[:layout] = layout if layout.present?
+      opts[:layout] = layout if layout.present? || layout == false
```
Nice one!
